### ISSUE TITLE
conda update --all broken in 4.4.8?

### ISSUE THREAD

**Issue by dhirschfeld** [conda/conda#6806]
**I'm submitting a...**
  - [ x ] bug report
  - [ ] feature request

## Current Behavior

Conda doesn't recognise jupyterlab as being upgradable when using `--all`
```
λ conda update --all
Solving environment: done

# All requested packages already installed.
```
...but is happy to upgrade it when asked to do so manually:
```
λ conda update jupyterlab
Solving environment: done

## Package Plan ##

  environment location: C:\Miniconda3

  added / updated specs:
    - jupyterlab


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    jupyterlab-0.31.2          |           py36_0        12.0 MB  https://conda.anaconda.org/conda-forge
    jupyterlab_launcher-0.10.3 |           py36_0          25 KB  https://conda.anaconda.org/conda-forge
    ------------------------------------------------------------
                                           Total:        12.0 MB

The following packages will be UPDATED:

    expat:               2.2.5-hcc4222d_0                                         --> 2.2.5-vc14_0  https://conda.anaconda.org/conda-forge [vc14]
    hdf5:                1.10.1-h98b8871_1                                        --> 1.10.1-vc14_1 https://conda.anaconda.org/conda-forge [vc14]
    jupyterlab:          0.30.2-py36_0     https://conda.anaconda.org/conda-forge --> 0.31.2-py36_0 https://conda.anaconda.org/conda-forge
    jupyterlab_launcher: 0.6.0-py36_0      https://conda.anaconda.org/conda-forge --> 0.10.3-py36_0 https://conda.anaconda.org/conda-forge
    libpng:              1.6.34-h79bbb47_0                                        --> 1.6.34-vc14_0 https://conda.anaconda.org/conda-forge [vc14]
    libtiff:             4.0.9-h0f13578_0                                         --> 4.0.9-vc14_0  https://conda.anaconda.org/conda-forge [vc14]
    libxml2:             2.9.7-h79bbb47_0                                         --> 2.9.7-vc14_0  https://conda.anaconda.org/conda-forge [vc14]
    libxslt:             1.1.32-hf6f1972_0                                        --> 1.1.32-vc14_0 https://conda.anaconda.org/conda-forge [vc14]

The following packages will be DOWNGRADED:

    curl:                7.55.1-h3b839b5_4                                        --> 7.55.1-vc14_0 https://conda.anaconda.org/conda-forge [vc14]
    gflags:              2.2.1-h6538335_0                                         --> 2.2.0-vc14_1  https://conda.anaconda.org/conda-forge [vc14]
    hdf4:                4.2.13-h712560f_2                                        --> 4.2.13-vc14_0 https://conda.anaconda.org/conda-forge [vc14]
    icu:                 58.2-ha66f8fd_1                                          --> 58.2-vc14_0   https://conda.anaconda.org/conda-forge [vc14]
    krb5:                1.14.2-h63dfc2a_6                                        --> 1.14.2-vc14_0 https://conda.anaconda.org/conda-forge [vc14]
    libnetcdf:           4.5.0-h1f49dca_7                                         --> 4.5.0-vc14_3  https://conda.anaconda.org/conda-forge [vc14]
    libssh2:             1.8.0-hd619d38_4                                         --> 1.8.0-vc14_2  https://conda.anaconda.org/conda-forge [vc14]
    lz4-c:               1.8.0-h047fa9f_1                                         --> 1.8.0-vc14_0  https://conda.anaconda.org/conda-forge [vc14]
    snappy:              1.1.7-h777316e_3                                         --> 1.1.7-vc14_1  https://conda.anaconda.org/conda-forge [vc14]
    sqlite:              3.21.0-h9d3ae62_0                                        --> 3.20.1-vc14_2 https://conda.anaconda.org/conda-forge [vc14]
    tk:                  8.6.7-hcb92d03_3                                         --> 8.6.7-vc14_0  https://conda.anaconda.org/conda-forge [vc14]
    yaml:                0.1.7-hc54c509_2                                         --> 0.1.7-vc14_0  https://conda.anaconda.org/conda-forge [vc14]

Proceed ([y]/n)?
```

**Comment by kalefranz** [comment]
`update --all` is really complex.  You can see when you updated jupyterlab manually, there we a BUNCH of downgrades required.  In fact, more downgrades then upgrades.  `update --all` doesn't know that jupyterlab happens to be the most important package to you.  So asking for the package update directly is the right solution.

**Comment by dhirschfeld** [comment]
@kalefranz - those aren't *upgrades* - they're just the package shuffling described in #6284 

After accepting the supposed *downgrades* I then did a `conda update --all` which then exactly undid all the downgrades, reverting the packages back to the defaults channel. After doing that a further `conda update --all` says I'm actually on the latest.

So the *downgrades* are in fact just shuffling the exact same version of packages between the defaults and the conda-forge channel.

I can accept that the solver doesn't  know that the downgrades aren't actually downgrades so it's doing something sensible according to it's view of the world but in practice it means an `upgrade --all` doesn't upgrade a package when it's perfectly possible to do so.

i.e. after running the below commands:
```
conda update jupyterlab
conda update--all
```
the state of my environment is exactly the same but with `jupyterlab=0.31.2` and `jupyterlab_launcher=0.10.3`.

If the end result is a valid environment a single `conda update--all` should have chosen the plan which upgraded those two packages and changed no other packages.

**Comment by kalefranz** [comment]
@mcg1969 hates `update —all` because it really is non-trivial to get right. Actually there probably isn’t a “right” way to do it in all cases. To make it always give the answer people expect. Give me a set of reproducible steps from a completely clean environment though, and I’ll take a look.

**Comment by kalefranz** [comment]
Preferably on macOS or Linux though, please :-/

**Comment by dhirschfeld** [comment]
I'll see if I can create a minimal repro...

**Comment by mcg1969** [comment]
Yeah, I don’t consider this incorrect. Those _are_ downgrades by conda’s measurement, because of channel priority. (And channel priority is a bigger downgrade than even version bumps.)

**Comment by mcg1969** [comment]
But by all means, we did find a solver bug this week. If you can demonstrate that conda is not doing what we intended we can fix it. But `conda update —all` is always going to be fraught with odd behavior.

**Comment by dhirschfeld** [comment]
@mcg1969 - I have channel priority to false as I specifically don't want it to take a part in the choosing of packages - all I care about is having the latest version from any of my specified channels.

Because all the versions are in fact the same a perfectly valid environment would be to simply not downgrade *any* packages.

The fact that the solver chooses a whole bunch of entirely unnecessary downgrades (channel shuffling) would seem to a bug in the solver? 

i.e. if `channel_priority=false` the solver should *never* choose to swap a package from one channel for the same version from another channel.

**Comment by mcg1969** [comment]
Even with channel priority set to false it has to order its package preference somehow.

But this actually looks like a feature issue may be involved.

**Comment by mcg1969** [comment]
Yeah this is a feature activation. I can’t study it now but let’s keep this flagged until we understand it.

I do not think this would happen with channel priority set to true, however.

**Comment by mcg1969** [comment]
Actually it might. Again, let’s leave this open.

**Comment by dhirschfeld** [comment]
> *Even with channel priority set to false it has to order its package preference somehow.*

I guess my complaint with the channel shuffling is that it chooses an unnecessary change in the installed packages.

To prevent unnecessary changes I think it would be good to consider whether or not the package is already installed - e.g. an already installed version will always beat a choice that isn't installed 

So for `channel_priority=false` the ordering would be:
```
(version_number, is_already_installed, channel_priority)
```

In this way `channel_priority` is only ever used to break a tie when a new package is being installed in all other cases the already installed package wins if there are no greater version numbers to choose from.

**Comment by gnomic** [comment]
I'm having the same problem; This is a mess. It all started when I added tensorflow or keras which was pulled from conda-forge. Now when I run an upgrade I get a rollback. 

Here is the error info:

```
conda-4.4.8-py 100% |###############################| Time: 0:00:00   3.13 MB/s
ERROR conda.core.link:_execute_actions(339): An error occurred while installing
package 'defaults::ipykernel-4.8.0-py36_0'.
LinkError: post-link script failed for package defaults::ipykernel-4.8.0-py36_0
running your command again with `-v` will provide additional information
location of failed script: C:\Users\212563417\AppData\Local\Continuum\Anaconda3\
Scripts\.ipykernel-post-link.bat
==> script messages <==
<None>

Attempting to roll back.


LinkError: post-link script failed for package defaults::ipykernel-4.8.0-py36_0
running your command again with `-v` will provide additional information
location of failed script: C:\Users\212563417\AppData\Local\Continuum\Anaconda3\
Scripts\.ipykernel-post-link.bat
==> script messages <==
<None>

Here is the command and output:
```

> C:\Users\12345678>conda update  --all --dry-run
> Fetching package metadata .............
> Solving package specifications: .
> 
> Package plan for installation in environment C:\Users\212563417\AppData\Local\Continuum\Anaconda3:
> 
> The following NEW packages will be INSTALLED:
> 
>     pathlib:             1.0.1-py36_0
> 
> The following packages will be UPDATED:
> 
>     anaconda-client:     1.6.6-py36ha174c20_0              --> 1.6.8-py36_0
>     anaconda-navigator:  1.6.11-py36h25c516c_0             --> 1.6.12-py36hdad2993_0
>     astroid:             1.5.3-py36h9d85297_0              --> 1.6.0-py36_0
>     attrs:               17.3.0-py36hc87868e_0             --> 17.4.0-py36_0
>     babel:               2.5.0-py36h35444c1_0              --> 2.5.3-py36_0
>     bzip2:               1.0.6-vc14_1          conda-forge [vc14] --> 1.0.6-haa5b126_2
>     certifi:             2017.11.5-py36_0      conda-forge --> 2018.1.18-py36_0
>     comtypes:            1.1.2-py36heb9b3d1_0              --> 1.1.4-py36_0
>     conda:               4.3.32-py36_0         conda-forge --> 4.4.8-py36_0
>     conda-build:         3.2.2-py36_0                      --> 3.3.0-py36_0
>     curl:                7.55.1-vc14_0         conda-forge [vc14] --> 7.55.1-h3b839b5_4
>     ftfy:                4.4.2-py36_0          conda-forge --> 4.4.3-py36hb6b3982_0
>     icu:                 58.2-vc14_0           conda-forge [vc14] --> 58.2-ha66f8fd_1
>     ipykernel:           4.7.0-py36h2f9c1c0_0              --> 4.8.0-py36_0
>     jupyter_client:      5.2.1-py36_0                      --> 5.2.2-py36_0
>     jupyterlab:          0.30.6-py36h047fa9f_0             --> 0.31.1-py36_0
>     jupyterlab_launcher: 0.6.0-py36_0                      --> 0.10.2-py36_0
>     krb5:                1.14.2-vc14_0         conda-forge [vc14] --> 1.14.2-h63dfc2a_6
>     libiconv:            1.14-vc14_4           conda-forge [vc14] --> 1.15-h1df5818_7
>     libssh2:             1.8.0-vc14_2          conda-forge [vc14] --> 1.8.0-hd619d38_4
>     libxml2:             2.9.5-vc14_1          conda-forge [vc14] --> 2.9.7-h79bbb47_0
>     lxml:                4.1.1-py36he0adb16_0              --> 4.1.1-py36hef2cd61_1
>     matplotlib:          2.1.1-py36h2062329_0              --> 2.1.2-py36h016c42a_0
>     msgpack-python:      0.4.8-py36_0          conda-forge --> 0.5.1-py36he980bc4_0
>     networkx:            2.0-py36hff991e3_0                --> 2.1-py36_0
>     notebook:            5.3.1-py36_0                      --> 5.3.1-py36_1
>     numpy:               1.13.3-py36ha320f96_0             --> 1.13.3-py36h4a99626_2
>     openpyxl:            2.4.9-py36h6c6577a_0              --> 2.4.10-py36_0
>     patsy:               0.4.1-py36h42cefec_0              --> 0.5.0-py36_0
>     pep8:                1.7.0-py36h0f3d67a_0              --> 1.7.1-py36_0
>     pycurl:              7.43.0-py36h086bf4c_3             --> 7.43.0.1-py36h74b6da3_0
>     pylint:              1.7.4-py36ha4e6ded_0              --> 1.8.1-py36_0
>     pyodbc:              4.0.21-py36h82c9817_0             --> 4.0.22-py36h6538335_0
>     pytest:              3.3.0-py36hf3522a2_0              --> 3.3.2-py36_0
>     pywin32:             221-py36h9c10281_0                --> 222-py36hfa6e2cd_0
>     qt:                  5.6.2-vc14_1          conda-forge [vc14] --> 5.6.2-vc14h6f8c307_12
>     scikit-image:        0.13.0-py36h6dffa3f_1             --> 0.13.1-py36hfa6e2cd_1
>     sphinx:              1.6.3-py36h9bb690b_0              --> 1.6.6-py36_0
>     spyder:              3.2.5-py36_0                      --> 3.2.6-py36_0
>     sqlalchemy:          1.1.13-py36h5948d12_0             --> 1.2.1-py36hfa6e2cd_0
>     sqlite:              3.20.1-vc14_2         conda-forge [vc14] --> 3.21.0-h9d3ae62_2
>     tk:                  8.6.7-vc14_0          conda-forge [vc14] --> 8.6.7-hcb92d03_3
>     xlwings:             0.11.4-py36hd3cf94d_0             --> 0.11.5-py36_0
>     yaml:                0.1.7-vc14_0          conda-forge [vc14] --> 0.1.7-hc54c509_2
>     zlib:                1.2.11-vc14_0         conda-forge [vc14] --> 1.2.11-h8395fce_2
> 
> The following packages will be SUPERSEDED by a higher-priority channel:
> 
>     ca-certificates:     2017.11.5-0           conda-forge --> 2017.08.26-h94faf87_0
>     cymem:               1.31.2-py36_vc14_0    conda-forge [vc14] --> 1.31.2-py36h51d26f2_0
>     cytoolz:             0.8.2-py36_0          conda-forge --> 0.8.2-py36h547e66e_0
>     dill:                0.2.7.1-py36_0        conda-forge --> 0.2.7.1-py36hf552773_0
>     freetype:            2.8.1-vc14_0          conda-forge [vc14] --> 2.8-h51f8f2c_1
>     hdf5:                1.10.1-vc14_1         conda-forge [vc14] --> 1.10.1-h98b8871_1
>     jpeg:                9b-vc14_2             conda-forge [vc14] --> 9b-hb83a4c4_2
>     libpng:              1.6.34-vc14_0         conda-forge [vc14] --> 1.6.34-h79bbb47_0
>     libtiff:             4.0.9-vc14_0          conda-forge [vc14] --> 4.0.9-h0f13578_0
>     libxslt:             1.1.32-vc14_0         conda-forge [vc14] --> 1.1.32-hf6f1972_0
>     msgpack-numpy:       0.4.1-py_0            conda-forge --> 0.4.1-py36h93564ae_0
>     murmurhash:          0.28.0-py36_vc14_0    conda-forge [vc14] --> 0.26.4-py36h2b517bc_0
>     openssl:             1.0.2n-vc14_0         conda-forge [vc14] --> 1.0.2n-h74b6da3_0
>     plac:                0.9.6-py36_0          conda-forge --> 0.9.6-py36_0
>     preshed:             1.0.0-py36_vc14_0     conda-forge [vc14] --> 1.0.0-py36h065ec1e_0
>     regex:               2017.11.09-py36_0     conda-forge --> 2017.11.09-py36ha090894_0
>     spacy:               2.0.4-py36_vc14_0     conda-forge [vc14] --> 1.9.0-py36_vc14hffcad6b_1
>     termcolor:           1.1.0-py36_1          conda-forge --> 1.1.0-py36_0
>     thinc:               6.10.0-py36_vc14_0    conda-forge [vc14] --> 6.5.2-py36_vc14he7005f8_0
>     tqdm:                4.19.5-py_0           conda-forge --> 4.19.4-py36h02a35f0_0
>     ujson:               1.35-py36_0           conda-forge --> 1.35-py36_0
>

**Comment by neilser** [comment]
Earlier today I tried and failed to get an ancient Anaconda installation to update itself. I then wiped it completely and installed 5.0.1.
It soon told me that 90+ packages were upgradeable. I'm a conda noob and I know of no other way to update lots of packages than to go to the shell and type "conda update --all". To my surprise this only updated around 30 packages. Repeated invocations of update --all then told me "All requested packages already installed", like the OP above. However, I can manually update 'em...

Now, I have no problem with there being a better way to update dozens of packages than to use "update --all". However, there doesn't appear to be one - hopefully I'm wrong about that though? I can't understand though why the navigator insists on telling me that 64 packages are upgradable, when it isn't capable of doing a mass-update by itself...

Much like the earlier posters, I can happily update individual packages. For example, I updated pandas to 0.22.0. I just tried "update --all" after the pandas update and it now wants to update these:
intel-openmp-2018.0.0
mkl-2018.0.1
scipy-1.0.0
numexpr-2.6.4
numpy-1.13.3

I believe my installation was TOTALLY vanilla before my first "update --all".

**Comment by kalefranz** [comment]
@neilser is running into https://github.com/ContinuumIO/anaconda-issues/issues/4298

**Comment by neilser** [comment]
Quite possible, @kalefranz - I don't fully understand what's going on.
I did attempt to update some more packages this morning from the navigator GUI and I failed!
First I tried a few (incl. ipython, matplotlib and numpy) and got some errors.
Tried other permutations, including just numpy on its own; still getting basically the same error:
```UnsatisfiableError: The following specifications were found to be in conflict:
  - blaze -> cytoolz -> python[version='>=3.6,<3.7.0a0'] -> vc=14
  - blaze -> h5py -> hdf5[version='>=1.10.1,<1.10.2.0a0'] -> *[track_features=vc14]
  - blaze -> pytables[version='>=3.0.0'] -> numexpr
  - numpy==1.14.0
Use "conda info <package>" to see the dependencies for each package.
```

Given that I started with as clean of a sheet as I could manage (I should note that after I uninstalled Anaconda it didn't vanish from "installed programs" list in Windows, despite the uninstaller removing the parent folder successfully, and I also had to manually remove it from my PATH) I have no clue how to proceed, but I do agree with the comments in the other issue that it shouldn't be necessary for people to resort to searching for complicated SO answers just to update their packages  :stuck_out_tongue_winking_eye:
